tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -12601,13 +12710,20 @@ bool SpirvEmitter::emitEntryFunctionWrapper(const FunctionDecl *decl,
 
   // Create temporary variables for holding function call arguments
   llvm::SmallVector<SpirvInstruction *, 4> params;
-  for (const auto *param : decl->params()) {
-    const auto paramType = param->getType();
+  for (auto *param : decl->params()) {
+    QualType paramType = param->getType();
+    if (param->hasAttr<HLSLNoInterpolationAttr>() ||
+        paramType->isStructureType()) {
```

Comment:
The function `redeclSpecialVarType` seems to have an odd interface, which make the code harder to read.  If the type is a struct, then you call `redeclSpecialVarType`, and it will check for the attribute.  If the type is not a struct and not an array, then it does not look for the attribute. The use of `redeclSpecialVarType` should be consistent. 

I think you should change `redeclSpecialVarType` it check for the attribute and remove this check.

Reply:
I'll move check point inside this function.